dart-mode.el

```diff
@@ -562,6 +579,251 @@ This uses `dart-format-path' to find the formatter."
+    (set-process-filter
+     process
+     `(lambda (proc string)
+        (dart-analysis-server-process-filter ',das-inst string)))
+    das-inst))
+
+(defun dart-add-analysis-server-change-hooks ()
+  "Report the dirty buffer state to Dart analysis server."
+  (add-hook 'first-change-hook 'dart-add-analysis-overlay t t)
+  (add-hook 'after-change-functions 'dart-change-analysis-overlay t t)
+  (add-hook 'after-save-hook 'dart-remove-analysis-overlay t t))
+
+(defun dart-add-analysis-overlay ()
+  "Report to Dart analysis server that it should overlay this buffer."
+  (dart-analysis-server-send
+   "analysis.updateContent"
+   `((files .
+            ((,buffer-file-name . ((type . "add")
+                                   (content . ,(buffer-string)))))))))
+
+(defun dart-change-analysis-overlay
+    (change-begin change-end change-before-length)
+  "Report to analysis server that it should change the overlay for this buffer.
+
+The region that changed ranges from CHANGE-BEGIN to CHANGE-END, and the
+length of the text before the change is CHANGE-BEFORE-LENGTH."
+  (combine-after-change-calls
+    (dart-analysis-server-send
+     "analysis.updateContent"
+     `((files
+        . ((,buffer-file-name
+            . ((type . "change")
+               (edits
+                . (((offset . ,change-begin)
+                    (length . ,change-before-length)
+                    (replacement
+                     . ,(buffer-substring change-begin change-end)))))))))))))
+
+(defun dart-remove-analysis-overlay ()
+  "Remove the overlay for the current buffer since it has been saved."
+  (dart-analysis-server-send
+   "analysis.updateContent"
+   `((files . ((,buffer-file-name . ((type . "remove"))))))))
+
+(defun dart-add-current-file-to-analysis-roots ()
+  "Add the current file's pub root or directory to analysis roots."
+  (let ((pub-root (locate-dominating-file buffer-file-name "pubspec.yaml"))
+        (current-dir (file-name-directory buffer-file-name)))
+    (if pub-root (dart-add-to-analysis-roots (expand-file-name pub-root))
+      (dart-add-to-analysis-roots (expand-file-name current-dir)))))
+
+(defun dart-add-to-analysis-roots (dir)
+  "Add DIR to the analysis roots."
+  (add-to-list 'dart-analysis-roots dir)
+  (dart-set-analysis-roots))
+
+(defun dart-set-analysis-roots ()
+  "Send the current list of analysis roots to the analysis server."
+  (dart-analysis-server-send
+   "analysis.setAnalysisRoots"
+   `(("included" . ,dart-analysis-roots)
+     ("excluded" . nil))))
+
+(defun dart-analysis-server-send (method &optional params callback)
+  "Send the METHOD request to the server with optional PARAMS.
+
+If you provide a CALLBACK, it will be called with the JSON decoded
+response.  Otherwise, the output will just be checked."
+  (let ((req-without-id (dart-analysis-server-make-request method params)))
+    (dart-analysis-server-enqueue req-without-id callback)))
+
+(defun dart-analysis-server-make-request (method &optional params)
```

Comment:
This and the following functions through `dart-analysis-server-handle-msg` should all be private.